core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java

```diff
@@ -227,7 +227,29 @@ void observe(
     );
 
     /**
-     * Resets the underlying DataStore to its pre-initialized state such that no data remains on the local
+     * Starts the DataStore.  This only needs to be called if you wish to start eagerly.  If you don't call it,
+     * it will be called automatically prior to executing any other operations (#query, #save, #delete, #observe).
+     *
+     * @param onComplete Invoked after DataStore is initialized.  This does not block until subscriptions and
+     *                  sync are complete.  To block until sync and subscriptions are complete, use
+     *                   Amplify.Hub.subscribe to listen for the DataStoreChannelEventName.READY event on
+     *                   HubChannel.DATASTORE.
+     * @param onError Invoked if an exception occurs.
```

Comment:
Similar thoughts as above, about the `onComplete`, here. But also: what's the DX for this callback, why would someone use it? Shouldn't `start(...)` wait until model synchronization is fully started, to invoke this? Otherwise, why wouldn't I just always use the `READY` event? And if I should always prefer `READY`, then couldn't we remove this callback? Lastly, we now have:
1. Configured
2. Initialized
3. Started
4. Ready

All of which mean different things. That might be okay _internally_ to our code, but I wonder if we can eliminate ~2 of these entirely from the user's view.

Reply:
I've struggled with this one. If the intent is to align with JS (as much as possible), `start()` really means `startSyncEngine()`, but the naming is besides the point here. One of the things we discussed is that for the mobile platforms, blocking execution during the entirety of the initialization of sync engine could lead to poor UX.

In the past, I've advocated for blocking only until `StorageObserver` starts since that's the bare minimum we would need to ensure no customer data is lost. That being said, that would give the onComplete callback yet another meaning, so I'm not too sure about that.

Looking at the plugin code, it looks like the `onComplete` callback basically lets them know that we've received their request to start the sync engine. I commented under the start method of the plugin as well, but there may be a scenario which both onComplete and onError can fire for the same call, which could lead to confusion.

Maybe the "easy" answer is to have no callbacks on start and stop, and tell devs to listen to sync engine events on the Hub.

Reply:
I've been scratching my head for a while about this as well :).

Here's how `start()` is implemented in this PR:
 - `onComplete` is called after:
   - plugin initialization completes (if needed)
   - `storageObserver.startObservingStorageChanges` completes.
- `onError` is called:
  - if initialization doesn't complete within 2 seconds
  - if orchestrator start stop lock is not acquired within 2 seconds
  - if `storageObserver.startObservingStorageChanges` doesn't complete within 2 seconds

Realistically, if `onError` is called, the error is not going to be something clients can handle anyway.  It'll probably only be called because of a bug in the library.

A basic app built with DataStore should look something like this (`start()` is not even a critical component):

```
private void showListOfTodos() {
    runQuery(); // Show locally cached data (or local synced data if already synced)
    Amplify.Hub.subscribe(HubChannel.DATASTORE,
            hubEvent -> DataStoreChannelEventName.READY.toString().equals(hubEvent.getName()),
            hubEvent -> runQuery()); // Show synced data
}

private void runQuery() {
    Amplify.DataStore.query(Todo.class,
            result -> { /* Display results on UI */},
            error -> { }
    );
}
```

Customers only need to call `start()` if they want to pre-cache data before displaying it to the end user.  They would call start(), presumably on app launch, and then call `query` whenever the end user navigates to a view that requires data.   Based on this, I don't think onComplete / onError make much sense, so I will remove them.  

For `stop()`, there are two customer use cases:
1. Close subscription sockets to save battery when DataStore is not needed, such as on app close.
2. Force a re-evaluation of sync query predicates. A customer would immediately call `query(...)` after `stop()`. Since we have the startStopSemaphore locking the stop operation, this should work just fine.

Callbacks aren't particularly useful for either of those use cases.   I'll bring this up with iOS/JS in the DataStore sync meeting tomorrow, and if we all agree, we will remove the callbacks.

Reply:
I spoke with the iOS and JS teams about this a few days ago, and they opposed removing these callbacks. 

The JS version of `start()` returns a Promise, which will throw an exception if there are any exceptions thrown while evaluating the client provided selection sync query code block (not implemented on Android yet, but coming soon)

So, there are reasons for having callbacks, and reasons not to.  Given that adding this new API is a bit of a one-way door, I'd like to keep the callbacks, even if they don't report very meaningful errors, since we will likely need them later (probably for selective sync). 

If there are no objections to the above, this PR is ready to fly, and I'll merge as soon as I get an approval!